Added command timeout

Vladislav-Kisliy opened the pull request:
improved a consuming performance, fixed an unit test, added device emulation to unit test

peterjhill commented:
Any chance you can squash this into less commits?

peterjhill commented on src/main/java/net/juniper/netconf/Device.java:
```diff
@@ -62,7 +62,8 @@
     private String hostName;
     private int port;
-    private int timeout;
+    private int connectionTimeout;
+    private int commandTimeout;
```
great call. Like this!

peterjhill commented on src/main/java/net/juniper/netconf/NetconfSession.java:
```diff
@@ -61,24 +63,29 @@
-    NetconfSession(Channel netconfChannel, int timeout, String hello,
+    NetconfSession(Channel netconfChannel, int connectionTimeout, int commandTimeout,
+                   String hello,
                    DocumentBuilder builder) throws IOException {
```
do you have an overloaded method to make one of these with the old signature? The github diff is not making it obvious.

It would be nice to have both, plus it would be nice to have a builder to make one of these, but that is for a future feature pr. :)

peterjhill commented on src/main/java/net/juniper/netconf/XML.java:
```diff
@@ -52,10 +54,10 @@ protected XML(Element active) {
-            st = st.substring(st.indexOf("\n") +1);
-        if (st.endsWith("\n")) {
-            st = st.substring(0,st.lastIndexOf("\n"));
+        if (st.startsWith(LF))
+            st = st.substring(st.indexOf(LF) + 1);
+        if (st.endsWith(LF)) {
+            st = st.substring(0, st.lastIndexOf(LF));
```
love this!

Vladislav-Kisliy commented:
> Any chance you can squash this into less commits?

Vladislav-Kisliy commented:
@peterjhill 
Actually I thought you can squash when you merge it but anyway I'll do it.
Thanks for your comments, I'll change my PR and notify you

peterjhill commented:
I think the only change I would want is to bump the version in the pom.xml from
    <version>2.0.1</version>

to perhaps 2.1.0 ?
with the new required arg in netconf session class, it could be a breaking change?

Vladislav-Kisliy commented:
@peterjhill Hi. Please have a look. I returned old constructor, modified builder and squashed several commits. For the rest of them I'm not sure it's even possible without recreation my fork from the scratch

peterjhill commented:
I pinged someone I know at Juniper about the PR. Hopefully they can take a look at it. THanks again Vlad!

ydnath commented on src/main/java/net/juniper/netconf/Device.java:
```diff
@@ -88,6 +89,8 @@ public Device(
             @NonNull String hostName,
             Integer port,
             Integer timeout,
```
Is this still required? Where is this set?

peterjhill replied:
I think it is down  here in NetconfSession:

 NetconfSession(Channel netconfChannel, int connectionTimeout, int commandTimeout,
If only timeout is set, then Vlad's change sets both timeout flavors to the value of timeout.

Vladislav-Kisliy replied:
yes, it's for compatibility with old API
